File: src/collection.js
'use strict';
// Per-router collection configuration (#105).
//
// Global settings supply the default poll intervals. Everything about HOW a
// collector delivers (stream vs poll) and WHETHER it runs at all is per-router,
// because a fleet is not uniform: a hAP ac2 acting as an access point has no
// routed traffic, no Kid Control devices and no wireless registrations, yet it
// was asked for the same ~17 concurrent streams as a 1 GB hAP ax3. The evidence
// in #104 points at concurrent open channels rather than data volume, so being
// able to switch a router to polling, or to turn a collector off entirely, is
// the lever that matters.
//
// Deliberately pure: no I/O. This is the single source of truth that index.js,
// routers.js and alertSessions.js all resolve through, so there is exactly one
// place holding the precedence rules and exactly one list of collectors.

const { POLL_BOUNDS } = require('./settings');

/**
 * The collector registry. Everything else in this feature derives from it: the
 * modal checkboxes, the null-collector stub, the client card map, the
 * diagnostics list and the tests. Adding a collector here should be the only
 * edit needed to bring it into the feature.
 *
 *   key           identifier used in `off` and in the resolved maps
 *   sessionProp   property name on the session object built by buildSession()
 *   pollKey       settings.json interval key, or null when it has no global one
 *   defaultPollMs fallback used when pollKey is null
 *   streamKey     per-router override key; null when the collector never streams
 *   pollable      whether a poll path exists (or is being built) for it
 *   disableable   whether the user may turn it off
 *   requires      collectors whose data it cannot work without
 *   cards         dashboard card ids, so the client can mark them
 */
// `page` is the page whose view this collector feeds — the single source for the
// collector↔page edge (issue #108). null means the collector belongs to no one
// page: traffic and system drive the header gauges on every page, and arp emits
// nothing at all, it only feeds other collectors' name resolution. A collector
// that also surfaces as a dashboard card says so through `cards`, not by
// claiming a second page.
const COLLECTORS = Object.freeze([
  // ── Protected: read directly by other collectors, or feed stored history ───
  { key: 'traffic', label: 'Traffic',      sessionProp: 'traffic',      pollKey: null,           defaultPollMs: 1000,
    streamKey: null,             pollable: false, disableable: false, requires: [], page: null, cards: ['trafficCard'] },
  { key: 'system', label: 'System / Gauges',       sessionProp: 'system',       pollKey: 'pollSystem',   defaultPollMs: 2000,
    streamKey: 'streamSystem',   pollable: true,  disableable: false, requires: [], page: null, cards: ['systemCard'] },
  { key: 'arp', label: 'ARP',          sessionProp: 'arp',          pollKey: 'pollArp',      defaultPollMs: 30000,
    streamKey: 'streamArp',      pollable: true,  disableable: false, requires: [], page: null, cards: [] },
  { key: 'dhcpLeases', label: 'DHCP Leases',   sessionProp: 'dhcpLeases',   pollKey: 'pollDhcp',     defaultPollMs: 600000,
    streamKey: 'streamLeases',   pollable: true,  disableable: false, requires: [], page: 'dhcp', cards: [] },
  { key: 'dhcpNetworks', label: 'DHCP Networks', sessionProp: 'dhcpNetworks', pollKey: 'pollDhcp',     defaultPollMs: 600000,
    streamKey: 'streamDhcp',     pollable: true,  disableable: false, requires: [], page: 'dhcp', cards: ['networksCard'] },

  // ── Disableable ────────────────────────────────────────────────────────────
  { key: 'conns', label: 'Connections',        sessionProp: 'conns',        pollKey: 'pollConns',    defaultPollMs: 5000,
    streamKey: 'streamConns',    pollable: true,  disableable: true,  requires: [], page: 'connections', cards: ['connCard'] },
  { key: 'bandwidth', label: 'Bandwidth',    sessionProp: 'bandwidth',    pollKey: 'pollBandwidth', defaultPollMs: 5000,
    streamKey: null,             pollable: true,  disableable: true,  requires: ['conns'], page: 'bandwidth', cards: ['bandwidthCard'] },
  { key: 'talkers', label: 'Top Talkers',      sessionProp: 'talkers',      pollKey: 'pollTalkers',  defaultPollMs: 3000,
    streamKey: 'streamTalkers',  pollable: true,  disableable: true,  requires: [], page: 'dashboard', cards: ['talkersCard'] },
  { key: 'ifStatus', label: 'Interface Rates',     sessionProp: 'ifStatus',     pollKey: 'pollIfstatus', defaultPollMs: 5000,
    streamKey: 'streamIfrates',  pollable: true,  disableable: true,  requires: [], page: 'interfaces', cards: ['ifStatusCard'] },
  { key: 'ping', label: 'Ping',         sessionProp: 'ping',         pollKey: 'pollPing',     defaultPollMs: 5000,
    streamKey: 'streamPing',     pollable: true,  disableable: true,  requires: [], page: 'dashboard', cards: [] },
  { key: 'wireless', label: 'Wireless',     sessionProp: 'wireless',     pollKey: 'pollWireless', defaultPollMs: 30000,
    streamKey: 'streamWireless', pollable: true,  disableable: true,  requires: [], page: 'wireless', cards: ['wirelessCard'] },
  { key: 'vpn', label: 'VPN',          sessionProp: 'vpn',          pollKey: 'pollVpn',      defaultPollMs: 10000,
    streamKey: 'streamVpn',      pollable: true,  disableable: true,  requires: [], page: 'vpn', cards: ['vpnCard'] },
  { key: 'firewall', label: 'Firewall',     sessionProp: 'firewall',     pollKey: 'pollFirewall', defaultPollMs: 5000,
    streamKey: 'streamFirewall', pollable: true,  disableable: true,  requires: [], page: 'firewall', cards: ['firewallCard'] },
  { key: 'routing', label: 'Routing',      sessionProp: 'routing',      pollKey: 'pollRouting',  defaultPollMs: 10000,
    streamKey: 'streamRouting',  pollable: true,  disableable: true,  requires: [], page: 'routing',
    cards: ['routingProtoCard', 'routingBgpCard', 'routingPeersCard', 'routingRoutesCard'] },
  { key: 'netwatch', label: 'NetWatch',     sessionProp: 'netwatch',     pollKey: null,           defaultPollMs: 30000,
    streamKey: 'streamNetwatch', pollable: true,  disableable: true,  requires: [], page: 'dashboard', cards: ['netwatchCard'] },
  { key: 'topology', label: 'Network Topology', sessionProp: 'topology', pollKey: 'pollTopology', defaultPollMs: 30000,
    streamKey: 'streamTopology', pollable: true,  disableable: true,  requires: [], page: 'topology', cards: ['topologyCard'] },
  // requires: [] on purpose for vlans. It reads live rates out of ifStatus and
  // client counts out of dhcpLeases, but declaring those here would cascade into
  // a hard disable — turning off Interface Rates would blank the whole VLANs
  // page, when membership, trunk ports and client counts are all still there.
  // Degrade the rates, not the page.
  { key: 'vlans', label: 'VLANs',        sessionProp: 'vlans',        pollKey: 'pollVlans',    defaultPollMs: 5000,
    streamKey: 'streamVlans',    pollable: true,  disableable: true,  requires: [], page: 'vlans', cards: [] },
  { key: 'ppp',   label: 'PPP',          sessionProp: 'ppp',          pollKey: 'pollPpp',      defaultPollMs: 5000,
    streamKey: 'streamPpp',      pollable: true,  disableable: true,  requires: [], page: 'ppp',   cards: [] },
  // bridges borrows rates from ifStatus the way vlans does, and for the same
  // reason declares no requires: without Interface Rates a bridge still has
  // ports, STP roles and a host table worth showing.
  { key: 'bridges', label: 'Bridges',    sessionProp: 'bridges',      pollKey: 'pollBridges',  defaultPollMs: 5000,
    streamKey: 'streamBridges',  pollable: true,  disableable: true,  requires: [], page: 'bridges', cards: [] },
  { key: 'capsman', label: 'CAPsMAN',    sessionProp: 'capsman',      pollKey: 'pollCapsman',  defaultPollMs: 10000,
    streamKey: 'streamCapsman',  pollable: true,  disableable: true,  requires: [], page: 'capsman', cards: [] },
  // dns, packages and rosusers are streamKey: null ON PURPOSE, and they are the
  // only entries in this registry that are. RouterOS would accept /listen on both menus, so this
  // is a choice rather than a limitation:
  //
  //   dns       the settings row is one record and the static table is single
  //             digits, so there is nothing a channel would save. The expensive
  //             part is the cache, which is already opt-in and fetched only
  //             while somebody has the browser open — an open channel would
  //             hold a resource for a table nobody is looking at.
  //   packages  an inventory changes on a reboot, not on a tick. It polls every
  //             60 s and the page forces a refresh after an action, which is
  //             strictly better than a channel held open for weeks.
  //
  // Both still honour the router's poll interval, so the Poll/Stream switch has
  // nothing to change for them.
  { key: 'dns',   label: 'DNS',          sessionProp: 'dns',          pollKey: 'pollDns',      defaultPollMs: 10000,
    streamKey: null,             pollable: true,  disableable: true,  requires: [], page: 'dns',   cards: [] },
  { key: 'packages', label: 'Packages',  sessionProp: 'packages',     pollKey: 'pollPackages', defaultPollMs: 60000,
    streamKey: null,             pollable: true,  disableable: true,  requires: [], page: 'packages', cards: [] },
  // wan borrows rates from ifStatus the way vlans and bridges do, and declares
  // no requires for the same reason: switching Interface Rates off should cost
  // the rate column, not the page.
  { key: 'wan',   label: 'WAN',        sessionProp: 'wan',          pollKey: 'pollWan',      defaultPollMs: 10000,
    streamKey: 'streamWan',      pollable: true,  disableable: true,  requires: [], page: 'wan', cards: [] },
  // queues borrows the FastTrack summary from the firewall collector the way
  // vlans borrows rates from ifStatus, and declares no `requires` for the same
  // reason: a hard dependency would blank the whole Queues page when somebody
  // switched Firewall collection off. Degrade the banner, not the page.
  { key: 'queues', label: 'Queues',   sessionProp: 'queues',       pollKey: 'pollQueues',   defaultPollMs: 5000,
    streamKey: 'streamQueues',   pollable: true,  disableable: true,  requires: [], page: 'queues', cards: [] },
  // rosusers is the third streamKey: null entry, for the same reason as
  // packages rather than dns: a router's user list changes when an operator
  // edits it, which is a human-timescale event. It polls slowly and the page
  // forces a re-read after every action, so a channel held open for weeks would
  // buy nothing.
  { key: 'rosusers', label: 'Router Users', sessionProp: 'rosusers',  pollKey: 'pollRosusers', defaultPollMs: 60000,
    streamKey: null,             pollable: true,  disableable: true,  requires: [], page: 'rosusers', cards: [] },
  // logs stays streamed even in poll mode: /log/listen pushes new entries, and
  // polling /log/print would drop lines between polls. Correctness, not fidelity.
  { key: 'logs', label: 'Logs',         sessionProp: 'logs',         pollKey: null,           defaultPollMs: 0,
    streamKey: null,             pollable: false, disableable: true,  requires: [], page: 'logs', cards: [] },
]);

const BY_KEY      = Object.freeze(Object.fromEntries(COLLECTORS.map(c => [c.key, c])));
const DISABLEABLE = Object.freeze(COLLECTORS.filter(c => c.disableable).map(c => c.key));
const POLL_KEYS   = Object.freeze([...new Set(COLLECTORS.map(c => c.pollKey).filter(Boolean))]);
const STREAM_KEYS = Object.freeze(COLLECTORS.map(c => c.streamKey).filter(Boolean));

const DEFAULT_MODE = 'stream';
const MODES = Object.freeze(['stream', 'poll']);

/** Clamp an interval using the bounds settings.js already enforces. */
function clampPollValue(key, raw) {
  const n = Number.isFinite(Number(raw)) ? Math.trunc(Number(raw)) : NaN;
  if (!Number.isFinite(n)) return null;
  const bounds = POLL_BOUNDS && POLL_BOUNDS[key];
  if (!bounds) return n;
  return Math.max(bounds[0], Math.min(bounds[1], n));
}

/**
 * Resolve the effective collection config for one router.
 *
 * Precedence, lowest to highest:
 *   interval : global setting   <  router override
 *   delivery : router mode      <  router per-collector override
 *
 * Delivery takes NO global input by design: stream-vs-poll is a property of the
 * router, not of the installation. Mode switches delivery only and never touches
 * intervals, so choosing Poll cannot silently also mean "slower" — which would
 * be unrecoverable from the UI.
 */
// Legacy app-global stream-vs-poll keys, retired in #105. Read only by
// planMigration below; nothing else in the codebase may consult them.
const LEGACY_STREAM_KEYS = Object.freeze({
  streamSystem:  'system',
  streamPing:    'ping',
  streamConns:   'conns',
  streamTalkers: 'talkers',
  streamIfrates: 'ifStatus',
});

/**
 * Work out how the retired global Collection Method maps onto per-router blocks.
 *
 * Pure: returns `[{ id, collection }]` for the routers that need writing, so the
 * caller owns persistence and the mapping itself is testable. Routers that
 * already carry a `mode` are left alone — an explicit per-router choice must
 * never be overwritten by a global that no longer has a UI.
 *
 * All-true is the new default, so it needs no block at all; that is why the
 * common case returns an empty array and writes nothing.
 */
function planMigration(settings, routers) {
  const cfg     = settings || {};
  const present = Object.keys(LEGACY_STREAM_KEYS).filter(k => typeof cfg[k] === 'boolean');
  if (!present.length) return [];

  const allOff = present.every(k => cfg[k] === false);
  const allOn  = present.every(k => cfg[k] === true);
  if (allOn) return [];                       // already the default

  const plan = [];
  for (const r of (routers || [])) {
    if (r && r.collection && r.collection.mode) continue;
    if (allOff) {
      plan.push({ id: r.id, collection: { mode: 'poll' } });
    } else {
      // Mixed: stay on stream and record only the collectors that were polled.
      const overrides = {};
      for (const k of present) if (cfg[k] === false) overrides[k] = false;
      plan.push({ id: r.id, collection: { mode: 'stream', overrides } });
    }
  }
  return plan;
}

function resolveCollection(settings, routerRecord) {
  const cfg  = settings || {};
  const coll = (routerRecord && routerRecord.collection) || {};
  const mode = MODES.includes(coll.mode) ? coll.mode : DEFAULT_MODE;
  const off  = Array.isArray(coll.off) ? coll.off : [];
  const ovr  = (coll.overrides && typeof coll.overrides === 'object') ? coll.overrides : {};

  const poll = {}, stream = {}, enabled = {};

  for (const c of COLLECTORS) {
    const globalVal = c.pollKey ? cfg[c.pollKey] : undefined;
    const raw = (c.pollKey && ovr[c.pollKey] !== undefined) ? ovr[c.pollKey]
              : (globalVal !== undefined ? globalVal : c.defaultPollMs);
    const clamped = c.pollKey ? clampPollValue(c.pollKey, raw) : Math.trunc(Number(raw) || 0);
    poll[c.key] = clamped === null ? c.defaultPollMs : clamped;

    if (!c.pollable) {
      stream[c.key] = true;                        // logs, traffic: stream is the only path
    } else if (c.streamKey && ovr[c.streamKey] !== undefined) {
      stream[c.key] = ovr[c.streamKey] === true || ovr[c.streamKey] === 'true';
    } else if (c.streamKey) {
      stream[c.key] = mode !== 'poll';
    } else {
      stream[c.key] = false;                       // bandwidth: timer-driven, never a stream
    }

    enabled[c.key] = c.disableable ? !off.includes(c.key) : true;
  }

  // pollIfaces is interfaceStatus's *metadata* interval, not a collector of its
  // own, so it has no registry row — but it is override-able like any other.
  const _ifaces = clampPollValue('pollIfaces',
    ovr.pollIfaces !== undefined ? ovr.pollIfaces : cfg.pollIfaces);
  poll.ifaces = _ifaces === null ? 60000 : _ifaces;

  // pingEnabled is a separate global kill switch and still wins.
  if (cfg.pingEnabled === false) enabled.ping = false;

  // Cascade dependencies here rather than in the UI, so a hand-edited
  // routers.json cannot produce a combination that silently breaks a card.
  // Bandwidth has no fetch of its own: it reads connTableCache, which only the
  // connections collector fills. Loop until stable so a chain would also settle.
  let changed = true;
  while (changed) {
    changed = false;
    for (const c of COLLECTORS) {
      if (!enabled[c.key]) continue;
      if (c.requires.some(dep => !enabled[dep])) { enabled[c.key] = false; changed = true; }
    }
  }

  // `overrides` is passed through so callers can tell an inherited value from a
  // pinned one — the settings live-patch must not drag a pinned router back to
  // the fleet default.
  return { mode, poll, stream, enabled, overrides: ovr };
}

/**
 * Stable string answering "would this router's session be built differently?".
 * Lets a router edit skip the session rebuild when nothing that matters changed,
 * so a label-only edit costs no reconnect. Key and array order are normalised so
 * a cosmetic re-save produces an identical fingerprint.
 */
function collectionFingerprint(settings, routerRecord) {
  const r = resolveCollection(settings, routerRecord);
  const pick = (obj) => Object.keys(obj).sort().map(k => k + '=' + obj[k]).join(',');
  const rec = routerRecord || {};
  const cfg = settings || {};
  return [
    'mode=' + r.mode,
    'poll:' + pick(r.poll),
    'stream:' + pick(r.stream),
    'enabled:' + pick(r.enabled),
    // Not part of `collection`, but they change how the session is built too.
    ['defaultIf', 'pingTarget'].map(k => k + '=' + (rec[k] || '')).join(','),
    ['topN', 'topTalkersN', 'maxConns', 'historyMinutes']
      .map(k => k + '=' + (cfg[k] === undefined ? '' : cfg[k])).join(','),
  ].join('|');
}

module.exports = {
  COLLECTORS, BY_KEY, DISABLEABLE, POLL_KEYS, STREAM_KEYS, MODES, DEFAULT_MODE,
  clampPollValue, resolveCollection, collectionFingerprint, planMigration, LEGACY_STREAM_KEYS,
};

File: internal/collection/collection.go
// Package collection resolves the effective collection config for one router —
// the port of `src/collection.js`'s `resolveCollection`.
//
// ── WHAT IT IS FOR ──────────────────────────────────────────────────────────
//
// A fleet is not uniform. A hAP ac2 acting as an access point has no routed
// traffic, no DHCP clients and no wireless registrations, yet without this it is
// asked for the same concurrent channels as a 1 GB hAP ax3 — and the documented
// bottleneck is concurrent API channels on the MikroTik, not data volume. So
// per-router intervals, per-router stream-vs-poll, and turning a collector off
// entirely are the levers that matter (#105).
//
// ── THE PORT HAS BEEN RUNNING WITHOUT IT ────────────────────────────────────
//
// `session.Session` and `routers.Pool` both pass `0` for every interval and
// construct every collector, so every router in a fleet is served identically.
// This package is the decision; WIRING it into those two is a separate step,
// because a collector that must not be constructed cannot be handled by skipping
// `Start()` — eleven of the live sixteen open their streams from the
// constructor, and the port's own collectors will need the same treatment.
//
// ── THE REGISTRY IS EMBEDDED, NOT TYPED ─────────────────────────────────────
//
// Everything derives from the collector registry: intervals, stream keys, the
// disableable set, the dependency edges. It is 27 rows of data, and a Go literal
// copied by hand is a transcription error waiting to happen — one wrong
// `pollable` and a collector silently loses its poll path.
//
// `collection_tables.json` is GENERATED from the live module by
// The collection corpus and embedded here, the same arrangement
// `internal/store` uses for `settings_tables.json`. `--check` fails when it
// drifts, so the registry cannot rot quietly.
//
// ── MUTATIONS (2026-08-25), eight of eight killed ───────────────────────────
//
//	the global interval beats the router override      1
//	a stream override rejects the STRING "true"        1
//	an unknown mode is honoured                        1
//	a non-pollable collector may poll                 29
//	pollIfaces takes the wrong default                27
//	the pingEnabled kill switch is ignored             2
//	the dependency cascade never runs                  2
//	the clamp bounds are not applied                   2
package collection

import (
	_ "embed"
	"encoding/json"
	"math"
	"slices"
	"sort"
	"strconv"
	"strings"
)

//go:embed collection_tables.json
var tablesJSON []byte

// Collector is one registry row, reduced to what a resolver consumes.
type Collector struct {
	Key string `json:"key"`
	// Label is the human name `GET /api/collectors` sends. Carried in the
	// generated tables rather than typed here, for the same reason as the rest of
	// the registry: it is the live app's own wording.
	Label         string `json:"label"`
	PollKey       string `json:"pollKey"`
	DefaultPollMs int    `json:"defaultPollMs"`
	StreamKey     string `json:"streamKey"`
	Pollable      bool   `json:"pollable"`
	Disableable   bool   `json:"disableable"`
	// EmptyKey names the payload list(s) whose emptiness means "nothing to
	// report here". Its PRESENCE, with Disableable, is what makes a collector
	// eligible for dormancy — the live filter is
	// `_COLLECTOR_DEFS.filter(c => c.emptyKey && c.disableable)`.
	//
	// Always a list here. The live value is a string for some collectors and an
	// array for others, and `payloadEmpty` immediately normalises; the generator
	// does it instead, and asserts the shape so a third form fails there rather
	// than arriving as an empty list that silently makes a collector ineligible.
	EmptyKey []string `json:"emptyKey"`
	// SessionProp is the property the collector hangs off on the live session
	// object. Carried for the same reason as Label: the same names, not a second
	// set typed here.
	SessionProp string   `json:"sessionProp"`
	Requires    []string `json:"requires"`
}

type tables struct {
	Registry   []Collector           `json:"registry"`
	PollBounds map[string][2]float64 `json:"pollBounds"`
}

var loaded = mustTables()

func mustTables() tables {
	var t tables
	if err := json.Unmarshal(tablesJSON, &t); err != nil {
		// A build-time asset that does not parse is not a runtime condition.
		panic("collection: collection_tables.json: " + err.Error())
	}
	return t
}

// Collectors is the registry, in the order the live module declares it.
func Collectors() []Collector { return loaded.Registry }

// The live defaults, spelled here because they are not registry rows.
const (
	defaultMode         = "stream"
	defaultPollIfacesMs = 60000
)

// Resolved is the effective config for one router.
type Resolved struct {
	Mode string
	// Poll is keyed by collector, plus `ifaces` — interfaceStatus's METADATA
	// interval, which has no registry row of its own but is override-able like
	// any other.
	Poll    map[string]int
	Stream  map[string]bool
	Enabled map[string]bool
	// Overrides is passed through so a caller can tell an INHERITED value from a
	// PINNED one. The live comment says why: a settings live-patch must not drag
	// a pinned router back to the fleet default.
	Overrides map[string]any
}

// Router is the record's `collection` block, already decoded.
type Router struct {
	Mode      string
	Off       []string
	Overrides map[string]any
}

// Resolve computes one router's effective config.
//
// Precedence, lowest to highest:
//
//	interval : global setting  <  router override
//	delivery : router mode     <  router per-collector override
//
// Delivery takes NO global input by design: stream-vs-poll is a property of the
// router, not of the installation. Mode switches delivery only and never touches
// intervals, so choosing Poll cannot silently also mean "slower" — which would
// be unrecoverable from the UI.
func Resolve(settings map[string]any, r *Router) Resolved {
	mode := defaultMode
	var off []string
	ovr := map[string]any{}
	if r != nil {
		// AN UNKNOWN MODE IS NOT HONOURED. `MODES.includes(coll.mode)` on the
		// live side; a hand-edited routers.json saying "sideways" gets the
		// default rather than a third behaviour nothing implements.
		if r.Mode == "stream" || r.Mode == "poll" {
			mode = r.Mode
		}
		off = r.Off
		if r.Overrides != nil {
			ovr = r.Overrides
		}
	}

	out := Resolved{
		Mode:      mode,
		Poll:      map[string]int{},
		Stream:    map[string]bool{},
		Enabled:   map[string]bool{},
		Overrides: ovr,
	}

	for _, c := range loaded.Registry {
		// INTERVAL: override, then global, then the collector's own default.
		var raw any
		if c.PollKey != "" {
			if v, ok := ovr[c.PollKey]; ok {
				raw = v
			} else if v, ok := settings[c.PollKey]; ok {
				raw = v
			} else {
				raw = c.DefaultPollMs
			}
		} else {
			raw = c.DefaultPollMs
		}
		if c.PollKey != "" {
			if n, ok := clamp(c.PollKey, raw); ok {
				out.Poll[c.Key] = n
			} else {
				// `clampPollValue` returns null for anything non-finite, and the
				// live line reads `clamped === null ? c.defaultPollMs : clamped`.
				out.Poll[c.Key] = c.DefaultPollMs
			}
		} else {
			// No pollKey: `Math.trunc(Number(raw) || 0)`, with no bounds.
			out.Poll[c.Key] = trunc(raw)
		}

		// DELIVERY.
		switch {
		case !c.Pollable:
			// logs, traffic: a stream is the only path they have.
			out.Stream[c.Key] = true
		case c.StreamKey != "":
			if v, ok := ovr[c.StreamKey]; ok {
				// THE STRING FORM IS ACCEPTED TOO. The live test is
				// `=== true || === 'true'`, because a form submits strings; a
				// port reading only the boolean would ignore a real setting.
				out.Stream[c.Key] = v == true || v == "true"
			} else {
				out.Stream[c.Key] = mode != "poll"
			}
		default:
			// bandwidth: timer-driven, never a stream.
			out.Stream[c.Key] = false
		}

		out.Enabled[c.Key] = !c.Disableable || !slices.Contains(off, c.Key)
	}

	// `pollIfaces` is interfaceStatus's metadata interval — override-able, but
	// with its own default rather than a registry row.
	var ifRaw any
	if v, ok := ovr["pollIfaces"]; ok {
		ifRaw = v
	} else if v, ok := settings["pollIfaces"]; ok {
		ifRaw = v
	}
	if n, ok := clamp("pollIfaces", ifRaw); ok {
		out.Poll["ifaces"] = n
	} else {
		out.Poll["ifaces"] = defaultPollIfacesMs
	}

	// A separate GLOBAL kill switch, applied after the per-router off list and
	// still winning over it.
	if v, ok := settings["pingEnabled"]; ok && v == false {
		out.Enabled["ping"] = false
	}

	// DEPENDENCIES CASCADE, in a loop.
	//
	// Done here rather than in the UI so a hand-edited routers.json cannot
	// produce a combination that silently breaks a card: bandwidth has no fetch
	// of its own and reads the table only `conns` fills.
	//
	// The loop matters even though today's registry is one edge deep — see
	// The collection corpus, which records that a single pass would pass
	// every case, and that a second edge is a one-line registry change away.
	for changed := true; changed; {
		changed = false
		for _, c := range loaded.Registry {
			if !out.Enabled[c.Key] {
				continue
			}
			for _, dep := range c.Requires {
				if !out.Enabled[dep] {
					out.Enabled[c.Key] = false
					changed = true
					break
				}
			}
		}
	}
	return out
}

// clamp is `clampPollValue`: a non-finite input has NO value (the caller
// substitutes a default), and a known key is bounded.
func clamp(key string, raw any) (int, bool) {
	f, ok := num(raw)
	if !ok {
		return 0, false
	}
	n := math.Trunc(f)
	if b, has := loaded.PollBounds[key]; has {
		n = math.Max(b[0], math.Min(b[1], n))
	}
	return int(n), true
}

// trunc is `Math.trunc(Number(raw) || 0)` — the no-bounds path.
func trunc(raw any) int {
	f, ok := num(raw)
	if !ok {
		return 0
	}
	return int(math.Trunc(f))
}

// num is JavaScript's `Number(x)` for the shapes a settings file holds. A JSON
// decode gives float64 for every number, so that is the case that matters; the
// others are here because a caller may hand over a typed value.
func num(raw any) (float64, bool) {
	switch v := raw.(type) {
	case float64:
		if math.IsNaN(v) || math.IsInf(v, 0) {
			return 0, false
		}
		return v, true
	case int:
		return float64(v), true
	case json.Number:
		f, err := v.Float64()
		if err != nil {
			return 0, false
		}
		return f, true
	default:
		return 0, false
	}
}

// ParseRouter decodes a router record's `collection` block.
//
// IT NEVER FAILS, by design. The block is operator-editable and the live side
// tolerates rubbish in it — a non-array `off` and a non-object `overrides` are
// both silently ignored rather than honoured — so this mirrors that: anything it
// cannot make sense of is dropped, and a nil result resolves to the fleet
// defaults.
//
// The alternative, returning an error, would push the caller into deciding what
// to do with a router whose config is malformed, and the original has already
// decided: ignore the field, serve the router.
func ParseRouter(raw []byte) *Router {
	if len(raw) == 0 {
		return nil
	}
	var doc map[string]any
	if err := json.Unmarshal(raw, &doc); err != nil {
		// Not an object at all (`"collection": "yes"`, or malformed JSON).
		return nil
	}
	r := &Router{}
	if s, ok := doc["mode"].(string); ok {
		r.Mode = s
	}
	// `off` must be an ARRAY; a bare string names no collector even when it
	// spells one, because the live `includes` runs on an array.
	if list, ok := doc["off"].([]any); ok {
		for _, v := range list {
			if s, ok := v.(string); ok {
				r.Off = append(r.Off, s)
			}
		}
	}
	if m, ok := doc["overrides"].(map[string]any); ok {
		r.Overrides = m
	}
	return r
}

// Fingerprint answers "would this router's session be built differently?".
//
// A router edit rebuilds its session, and a rebuild is a reconnect. Comparing
// fingerprints lets a LABEL-ONLY edit cost nothing, which is what the live
// `collectionFingerprint` exists for. Key and slice order are normalised so a
// cosmetic re-save produces an identical string.
//
// ── IT FOLDS IN SIX FIELDS THAT ARE NOT PART OF `collection` ────────────────
//
// Two from the record (`defaultIf`, `pingTarget`) and four from settings
// (`topN`, `topTalkersN`, `maxConns`, `historyMinutes`). None of them changes
// what `Resolve` returns, and all of them change how the session is built — so
// two routers that resolve identically can still need different sessions.
//
// ── THE TWO GROUPS COERCE DIFFERENTLY, AND THAT IS NOT A TYPO ───────────────
//
// The record fields are `rec[k] || ”`, so `0` and `""` both render empty. The
// settings are `cfg[k] === undefined ? ” : cfg[k]`, so a `0` renders as `0`. A
// port that spelled both the same way would agree on every value except zero —
// and zero is a real setting for `topN`.
func Fingerprint(settings map[string]any, r *Router, rec RecordExtras) string {
	res := Resolve(settings, r)

	pickInt := func(m map[string]int) string {
		keys := make([]string, 0, len(m))
		for k := range m {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		parts := make([]string, 0, len(keys))
		for _, k := range keys {
			parts = append(parts, k+"="+strconv.Itoa(m[k]))
		}
		return strings.Join(parts, ",")
	}
	pickBool := func(m map[string]bool) string {
		keys := make([]string, 0, len(m))
		for k := range m {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		parts := make([]string, 0, len(keys))
		for _, k := range keys {
			parts = append(parts, k+"="+strconv.FormatBool(m[k]))
		}
		return strings.Join(parts, ",")
	}

	// `rec[k] || ''` — falsy becomes empty, which is why an empty string and a
	// zero are indistinguishable here.
	recPart := "defaultIf=" + truthy(rec.DefaultIf) + ",pingTarget=" + truthy(rec.PingTarget)

	// `cfg[k] === undefined ? '' : cfg[k]` — ABSENT becomes empty, present keeps
	// its value, including zero and false.
	cfgKeys := []string{"topN", "topTalkersN", "maxConns", "historyMinutes"}
	cfgParts := make([]string, 0, len(cfgKeys))
	for _, k := range cfgKeys {
		v, ok := settings[k]
		if !ok {
			cfgParts = append(cfgParts, k+"=")
			continue
		}
		cfgParts = append(cfgParts, k+"="+jsString(v))
	}

	return strings.Join([]string{
		"mode=" + res.Mode,
		"poll:" + pickInt(res.Poll),
		"stream:" + pickBool(res.Stream),
		"enabled:" + pickBool(res.Enabled),
		recPart,
		strings.Join(cfgParts, ","),
	}, "|")
}

// RecordExtras are the router-record fields the fingerprint folds in that are
// not part of the `collection` block.
type RecordExtras struct {
	DefaultIf  any
	PingTarget any
}

// truthy is JavaScript's `x || ”` for the shapes a record holds: anything falsy
// renders empty.
func truthy(v any) string {
	switch x := v.(type) {
	case nil:
		return ""
	case string:
		return x
	case bool:
		if !x {
			return ""
		}
		return "true"
	case float64:
		if x == 0 {
			return ""
		}
		return jsString(x)
	case int:
		if x == 0 {
			return ""
		}
		return strconv.Itoa(x)
	default:
		return jsString(v)
	}
}

// jsString renders a value the way JavaScript's string concatenation does. A
// JSON decode gives float64 for every number, and `1500 + ”` is "1500" rather
// than "1500.0" — so an integral float must lose its fraction.
func jsString(v any) string {
	switch x := v.(type) {
	case nil:
		return ""
	case string:
		return x
	case bool:
		return strconv.FormatBool(x)
	case float64:
		return strconv.FormatFloat(x, 'g', -1, 64)
	case int:
		return strconv.Itoa(x)
	default:
		return ""
	}
}
